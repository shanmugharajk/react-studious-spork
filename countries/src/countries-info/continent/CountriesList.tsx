import React from "react";

import { Alert, Info } from "~/components";
import { ICountry } from "~/countries-info/types";
import { isNull } from "~/utils";

interface IProps {
  countries?: ICountry[] | null;
  error?: string;
  loading: boolean;
}

export const CountriesList: React.FunctionComponent<IProps> = ({
  countries,
  error,
  loading,
}) => {
  if (error) {
    return <Alert>{error}</Alert>;
  }

  if (loading) {
    return <Info>Loading....</Info>;
  }

  // result from api will set this as null.
  if (isNull(countries)) {
    return <Info>No results found</Info>;
  }

  // We clear when user changes the input with undefined to immediately clear the
  // previous details shown to avoid confusion.
  if (!countries) {
    return null;
  }

  return (
    <ul className="mt-5 p-4 space-y-3 text-gray-600 text-sm h-64 rounded-lg border border-gray-300 overflow-auto">
      {countries.map((country) => (
        <li key={country.code}>
          <span className="inline-block mr-5">{country.emoji}</span>
          <span>{country.name}</span>
        </li>
      ))}
    </ul>
  );
};
