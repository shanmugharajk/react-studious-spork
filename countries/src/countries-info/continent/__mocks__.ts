import { GET_COUNTRIES_CONTINENT_CODE } from "./queries";

export const continentMockData = {
  name: "Asia",
  countries: [
    { code: "AM", name: "Armenia", emoji: "ðŸ‡¦ðŸ‡²" },
    { code: "BH", name: "Bahrain", emoji: "ðŸ‡§ðŸ‡­" },
    { code: "BN", name: "Brunei", emoji: "ðŸ‡§ðŸ‡³" },
    { code: "BT", name: "Bhutan", emoji: "ðŸ‡§ðŸ‡¹" },
    { code: "LA", name: "Laos", emoji: "ðŸ‡±ðŸ‡¦" },
    { code: "LB", name: "Lebanon", emoji: "ðŸ‡±ðŸ‡§" },
    { code: "MO", name: "Macao", emoji: "ðŸ‡²ðŸ‡´" },
    { code: "NP", name: "Nepal", emoji: "ðŸ‡³ðŸ‡µ" },
    { code: "OM", name: "Oman", emoji: "ðŸ‡´ðŸ‡²" },
    { code: "QA", name: "Qatar", emoji: "ðŸ‡¶ðŸ‡¦" },
    { code: "TR", name: "Turkey", emoji: "ðŸ‡¹ðŸ‡·" },
    { code: "TW", name: "Taiwan", emoji: "ðŸ‡¹ðŸ‡¼" },
    { code: "VN", name: "Vietnam", emoji: "ðŸ‡»ðŸ‡³" },
    { code: "YE", name: "Yemen", emoji: "ðŸ‡¾ðŸ‡ª" },
  ],
};

export const getCountriesContinentCodeMock = [
  {
    request: {
      query: GET_COUNTRIES_CONTINENT_CODE,
      variables: {
        code: "AS",
      },
    },
    result: {
      data: { continent: continentMockData },
    },
  },
];

export const getCountriesContinentCodeNoResultsMock = [
  {
    request: {
      query: GET_COUNTRIES_CONTINENT_CODE,
      variables: {
        code: "P",
      },
    },
    result: {
      data: { continent: null },
    },
  },
];

export const getCountriesContinentCodeErrorMock = [
  {
    request: {
      query: GET_COUNTRIES_CONTINENT_CODE,
      variables: {
        code: "AS",
      },
    },
    error: new Error("An error occurred"),
  },
];
