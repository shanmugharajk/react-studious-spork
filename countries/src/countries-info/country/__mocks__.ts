import { COUNTRIES_INFO_BY_ID } from "./queries";

export const countryMockData = {
  name: "India",
  code: "IN",
  currency: "INR",
  languages: [{ name: "Hindi" }, { name: "English" }],
};

export const getCountryInfoByIdMock = [
  {
    request: {
      query: COUNTRIES_INFO_BY_ID,
      variables: {
        code: "IN",
      },
    },
    result: {
      data: { country: countryMockData },
    },
  },
];

export const getCountryInfoByIdNoResultsMock = [
  {
    request: {
      query: COUNTRIES_INFO_BY_ID,
      variables: {
        code: "KU",
      },
    },
    result: {
      data: { country: null },
    },
  },
];

export const getCountryInfoByIdErrorMock = [
  {
    request: {
      query: COUNTRIES_INFO_BY_ID,
      variables: {
        code: "JU",
      },
    },
    error: new Error("An error occurred"),
  },
];
