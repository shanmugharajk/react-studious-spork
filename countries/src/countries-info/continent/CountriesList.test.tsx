import { render } from "@testing-library/react";

import type { ICountry } from "../types";
import { CountriesList } from "./CountriesList";

import { continentMockData } from "./__mocks__";

describe("CountriesList", () => {
  it("should render CountriesList", () => {
    const { container } = render(
      <CountriesList countries={continentMockData.countries as ICountry[]} />,
    );

    expect(container).toMatchSnapshot();
  });

  it("should render 'loading...' message", () => {
    const { container } = render(<CountriesList loading />);

    expect(container).toMatchSnapshot();
  });

  it("should render error message", () => {
    const { container } = render(<CountriesList error="An error occurred" />);

    expect(container).toMatchSnapshot();
  });

  it("should render 'No results found' message", () => {
    const { container } = render(<CountriesList countries={null} />);

    expect(container).toMatchSnapshot();
  });
});
