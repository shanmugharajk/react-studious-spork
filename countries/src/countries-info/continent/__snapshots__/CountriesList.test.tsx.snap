// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountriesList should render 'No results found' message 1`] = `
<div>
  <div
    class="mt-5 p-5 bg-gray-50 border border-gray-300 rounded-lg"
  >
    <span
      class="text-gray-500"
    >
      No results found
    </span>
  </div>
</div>
`;

exports[`CountriesList should render 'loading...' message 1`] = `
<div>
  <div
    class="mt-5 p-5 bg-gray-50 border border-gray-300 rounded-lg"
  >
    <span
      class="text-gray-500"
    >
      Loading....
    </span>
  </div>
</div>
`;

exports[`CountriesList should render CountriesList 1`] = `
<div>
  <ul
    class="mt-5 p-4 space-y-3 text-gray-600 text-sm h-64 rounded-lg border border-gray-300 overflow-auto"
  >
    <li
      data-testid="country-list-item"
    >
      <span
        class="inline-block mr-5"
      >
        ğŸ‡¦ğŸ‡²
      </span>
      <span>
        Armenia
      </span>
    </li>
    <li
      data-testid="country-list-item"
    >
      <span
        class="inline-block mr-5"
      >
        ğŸ‡§ğŸ‡­
      </span>
      <span>
        Bahrain
      </span>
    </li>
    <li
      data-testid="country-list-item"
    >
      <span
        class="inline-block mr-5"
      >
        ğŸ‡§ğŸ‡³
      </span>
      <span>
        Brunei
      </span>
    </li>
    <li
      data-testid="country-list-item"
    >
      <span
        class="inline-block mr-5"
      >
        ğŸ‡§ğŸ‡¹
      </span>
      <span>
        Bhutan
      </span>
    </li>
    <li
      data-testid="country-list-item"
    >
      <span
        class="inline-block mr-5"
      >
        ğŸ‡±ğŸ‡¦
      </span>
      <span>
        Laos
      </span>
    </li>
    <li
      data-testid="country-list-item"
    >
      <span
        class="inline-block mr-5"
      >
        ğŸ‡±ğŸ‡§
      </span>
      <span>
        Lebanon
      </span>
    </li>
    <li
      data-testid="country-list-item"
    >
      <span
        class="inline-block mr-5"
      >
        ğŸ‡²ğŸ‡´
      </span>
      <span>
        Macao
      </span>
    </li>
    <li
      data-testid="country-list-item"
    >
      <span
        class="inline-block mr-5"
      >
        ğŸ‡³ğŸ‡µ
      </span>
      <span>
        Nepal
      </span>
    </li>
    <li
      data-testid="country-list-item"
    >
      <span
        class="inline-block mr-5"
      >
        ğŸ‡´ğŸ‡²
      </span>
      <span>
        Oman
      </span>
    </li>
    <li
      data-testid="country-list-item"
    >
      <span
        class="inline-block mr-5"
      >
        ğŸ‡¶ğŸ‡¦
      </span>
      <span>
        Qatar
      </span>
    </li>
    <li
      data-testid="country-list-item"
    >
      <span
        class="inline-block mr-5"
      >
        ğŸ‡¹ğŸ‡·
      </span>
      <span>
        Turkey
      </span>
    </li>
    <li
      data-testid="country-list-item"
    >
      <span
        class="inline-block mr-5"
      >
        ğŸ‡¹ğŸ‡¼
      </span>
      <span>
        Taiwan
      </span>
    </li>
    <li
      data-testid="country-list-item"
    >
      <span
        class="inline-block mr-5"
      >
        ğŸ‡»ğŸ‡³
      </span>
      <span>
        Vietnam
      </span>
    </li>
    <li
      data-testid="country-list-item"
    >
      <span
        class="inline-block mr-5"
      >
        ğŸ‡¾ğŸ‡ª
      </span>
      <span>
        Yemen
      </span>
    </li>
  </ul>
</div>
`;

exports[`CountriesList should render error message 1`] = `
<div>
  <div
    class="mt-5 p-3 bg-red-50 border border-red-500 rounded-lg"
  >
    <span
      class="text-red-400"
    >
      An error occurred
    </span>
  </div>
</div>
`;
